go/libraries/doltcore/env/actions/checkout.go

```diff
@@ -92,3 +97,539 @@ func MoveTablesFromHeadToWorking(ctx context.Context, roots doltdb.Roots, tbls [
+	if err != nil {
+		return nil, err
+	}
+	return branchRoot, nil
+}
+
+// moveModifiedTables handles working set changes during a branch change.
+// When moving between branches, changes in the working set should travel with you.
+// Working set changes cannot be moved if the table differs between the old and new head,
+// in this case, we throw a conflict and error (as per Git).
+func moveModifiedTables(ctx context.Context, oldRoot, newRoot, changedRoot *doltdb.RootValue, conflicts *set.StrSet, force bool) (map[string]hash.Hash, error) {
+	resultMap := make(map[string]hash.Hash)
+	tblNames, err := newRoot.GetTableNames(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, tblName := range tblNames {
+		oldHash, _, err := oldRoot.GetTableHash(ctx, tblName)
+		if err != nil {
+			return nil, err
+		}
+
+		newHash, _, err := newRoot.GetTableHash(ctx, tblName)
+		if err != nil {
+			return nil, err
+		}
+
+		changedHash, _, err := changedRoot.GetTableHash(ctx, tblName)
+		if err != nil {
+			return nil, err
+		}
+
+		if oldHash == changedHash {
+			resultMap[tblName] = newHash
+		} else if oldHash == newHash {
+			resultMap[tblName] = changedHash
+		} else if newHash == changedHash {
+			resultMap[tblName] = oldHash
+		} else if force {
+			resultMap[tblName] = newHash
+		} else {
+			conflicts.Add(tblName)
+		}
+	}
+
+	tblNames, err = changedRoot.GetTableNames(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, tblName := range tblNames {
+		if _, exists := resultMap[tblName]; !exists {
+			oldHash, _, err := oldRoot.GetTableHash(ctx, tblName)
+			if err != nil {
+				return nil, err
+			}
+
+			changedHash, _, err := changedRoot.GetTableHash(ctx, tblName)
+			if err != nil {
+				return nil, err
+			}
+
+			if oldHash == emptyHash {
+				resultMap[tblName] = changedHash
+			} else if force {
+				resultMap[tblName] = oldHash
+			} else if oldHash != changedHash {
+				conflicts.Add(tblName)
+			}
+		}
+	}
+
+	return resultMap, nil
+}
+
+// moveForeignKeys returns the foreign key collection that should be used for the new working set.
+func moveForeignKeys(ctx context.Context, oldRoot, newRoot, changedRoot *doltdb.RootValue, force bool) (*doltdb.ForeignKeyCollection, error) {
+	oldFks, err := oldRoot.GetForeignKeyCollection(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	newFks, err := newRoot.GetForeignKeyCollection(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	changedFks, err := changedRoot.GetForeignKeyCollection(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	oldHash, err := oldFks.HashOf(ctx, oldRoot.VRW())
+	if err != nil {
+		return nil, err
+	}
+
+	newHash, err := newFks.HashOf(ctx, newRoot.VRW())
+	if err != nil {
+		return nil, err
+	}
+
+	changedHash, err := changedFks.HashOf(ctx, changedRoot.VRW())
+	if err != nil {
+		return nil, err
+	}
+
+	if oldHash == changedHash {
+		return newFks, nil
+	} else if oldHash == newHash {
+		return changedFks, nil
+	} else if newHash == changedHash {
+		return oldFks, nil
```

Comment:
This case didn't click with me... if the new root we're switching to has the same FK content as the working/staged root that was passed in, then we return the old FKs? From the earlier if case, we know that changedHash != oldHash. Seems odd to return the different `oldFks`.

Reply:
It's unreachable, good catch. Doesn't make sense, and would be a bug if it were reachable.